crypto/evp/pmeth_lib.c

```diff
@@ -265,7 +265,19 @@ static EVP_PKEY_CTX *int_ctx_new(OSSL_LIB_CTX *libctx,
      * fetching a provider implementation.
      */
     if (e == NULL && app_pmeth == NULL && keytype != NULL) {
-        keymgmt = EVP_KEYMGMT_fetch(libctx, keytype, propquery);
+        /*
+         * Prefer the keymgmt from the pkey if there is one. Otherwise fetch a
+         * new one.
+         */
+        if (pkey != NULL && pkey->keymgmt != NULL) {
+            if (!EVP_KEYMGMT_up_ref(pkey->keymgmt)) {
+                ERR_raise(ERR_LIB_EVP, ERR_R_INTERNAL_ERROR);
+                return NULL;
```

Comment:
EVP_PKEY_CTX_new_from_pkey() will now take keymgmt from pkey, instead of fetching one using propq.  The docs only say the id will be taken from pkey, leaving one to guess keymgmt will be fetched. It could be seen as a regression. Is this required for the patch to work?

Reply:
Not really. We *have* to use the key manager where the key is held at some point in the process, otherwise we cannot access the key. This just sets the initial "default" key manager to be the one in the pkey if it has one. When you actually use it for an operation we will fetch that operation algorithm using the propq. Subsequently we fetch a *new* key manager also using the propq (and ensuring that that new key manager is in the same provider as the operation algorithm, and export the key from the original key manager into the new one.

In the case where the original key manager is incapable of exporting the key, then all of the above still applies but with the additional restriction that when we fetch the algorithm operation we still use the same propq but *add* an extra requirement that the algorithm must come from the same provider as the existing key manager.

In this way the key manager/operation algorithm that eventually gets used should always be consistent with the propq.

Reply:
@mattcaswell, this is a semantic change, and if you really want this, then you absolutely need to modify the documentation for `EVP_PKEY_CTX_new_from_pkey()`, which currently says that the *keytype* of the given pkey is used (albeit a bit strangely written).

If this is becoming an indicator for what is to come, then I for one feel very uncertain how this would be interpretted:

``` C
 EVP_PKEY *pkey = load_key_from(NULL, "provider=myprovider");
 
 EVP_PKEY_CTX *ctx = EVP_PKEY_CTX_new_from_pkey(NULL, pkey, "provider=default");
```

Reply:
Hmmm, through my own coding, I'm starting to understand why this change came about.  I'm about to embark on the same myself...  which is a bit ironic, considering my protests here :wink: